File: Assets/src/model/tasks/MoveToTargetTask.cs
using System;
using UnityEngine;

[System.Serializable]
public class MoveToTargetTask : FollowPathTask {
  public MoveToTargetTask(Actor actor, Vector2Int target, Floor floor = null) : base(actor, target, (floor ?? GameModel.main.currentFloor).FindPath(actor.pos, target, false, actor)) {
  }
}

public class MoveToTargetThenPerformTask : MoveToTargetTask {
  private readonly Action then;

  public MoveToTargetThenPerformTask(Actor actor, Vector2Int target, Action then, Floor floor = null) : base(actor, target, floor) {
    this.then = then;
  }

  public override void PostStep(BaseAction action, BaseAction finalAction) {
    if (IsDone()) {
      then();
    }
  }
}

File: web/src/model/tasks/MoveToTargetTask.ts
import { FollowPathTask } from './FollowPathTask';
import { Vector2Int } from '../../core/Vector2Int';
import { GameModelRef } from '../GameModelRef';
import type { Actor } from '../Actor';
import type { BaseAction } from '../BaseAction';

export class MoveToTargetTask extends FollowPathTask {
  constructor(actor: Actor, target: Vector2Int) {
    const floor = GameModelRef.main.currentFloor;
    super(actor, target, floor.findPath(actor.pos, target, false, actor));
  }
}

export class MoveToTargetThenPerformTask extends MoveToTargetTask {
  private readonly then: () => void;

  constructor(actor: Actor, target: Vector2Int, then: () => void) {
    super(actor, target);
    this.then = then;
  }

  postStep(_action: BaseAction, _finalAction: BaseAction): void {
    if (this.isDone()) {
      this.then();
    }
  }
}
